internal/events/eventer.go

```diff
@@ -220,6 +216,25 @@ func recordMetrics(m messageInstruments) func(h message.HandlerFunc) message.Han
 	return metricsFunc
 }
 
+func poisonQueueTracking(ctx context.Context, instr *messageInstruments) func(h message.HandlerFunc) message.HandlerFunc {
+	metricsFunc := func(h message.HandlerFunc) message.HandlerFunc {
+		return func(msg *message.Message) ([]*message.Message, error) {
+			// Defer the tracking logic to after the message has been processed by other middlewares,
+			// including the deferred PoisonQueue middleware functionality,
+			// so that we can check if it has been poisoned or not.
+			defer func() {
+				if poisoned := msg.Metadata.Get(middleware.ReasonForPoisonedKey); poisoned != "" {
+					instr.poisonQueueCounter.Add(ctx, 1)
```

Comment:
I think this should use `msg.Context()`, not the outer context from when the metrics were set up.